python/tvm/topi/cuda/scatter_elements.py

```diff
@@ -0,0 +1,181 @@
+        If mul, the update values will be multiply to the input data
+        If min, there is choice of minimal between the update values and the input data
+        If max, there is choice of maximal between the update values and the input data
+        It is "update" by default
+
+    Returns
+    -------
+    ret : tvm.te.Tensor
+    """
+    if not isinstance(axis, int):
+        axis = get_const_int(axis)
+
+    def gen_ir(data, indices, updates, out, axis):
+        ib = tir.ir_builder.create()
+
+        data_ptr = ib.buffer_ptr(data)
+        indices_ptr = ib.buffer_ptr(indices)
+        updates_ptr = ib.buffer_ptr(updates)
+        out_ptr = ib.buffer_ptr(out)
+
+        # Prepare ranges and strides
+        shape = data.shape
+        if axis < 0:
+            axis = len(shape) + axis
+        axis_range = cast(shape[axis], indices.dtype)
+
+        before_axis_range = 1
+        after_axis_range = 1
+        for i, value in enumerate(shape, 0):
+            if i < axis:
+                before_axis_range *= value
+            elif i > axis:
+                after_axis_range *= value
+        before_axis_stride = axis_range * after_axis_range
+        full_range = before_axis_range * before_axis_stride
+
+        ind_shape = indices.shape
+        ind_axis_range = ind_shape[axis]
+
+        ind_before_axis_range = 1
+        ind_after_axis_range = 1
+        for i, value in enumerate(ind_shape, 0):
+            if i < axis:
+                ind_before_axis_range *= value
+            elif i > axis:
+                ind_after_axis_range *= value
+        ind_before_axis_stride = ind_axis_range * ind_after_axis_range
+        ind_full_range = ind_before_axis_range * ind_before_axis_stride
+        ind_full_range_excl_axis = ind_before_axis_range * ind_after_axis_range
+
+        max_threads = int(tvm.target.Target.current(allow_none=False).max_num_threads)
+        # Copy initial input data to output
+        with ib.new_scope():
+            num_blocks = ceil_div(full_range, max_threads)
+            bx = te.thread_axis("blockIdx.x")
+            tx = te.thread_axis("threadIdx.x")
+            ib.scope_attr(bx, "thread_extent", num_blocks)
+            ib.scope_attr(tx, "thread_extent", max_threads)
+
+            index = bx * max_threads + tx
+            with ib.if_scope(index < full_range):
+                out_ptr[index] = data_ptr[index]
+
+        # Check indices and shift to positive side if need
+        with ib.new_scope():
+            num_blocks_1 = ceil_div(ind_full_range, max_threads)
+            bx1 = te.thread_axis("blockIdx.x")
+            tx1 = te.thread_axis("threadIdx.x")
+            ib.scope_attr(bx1, "thread_extent", num_blocks_1)
+            ib.scope_attr(tx1, "thread_extent", max_threads)
+
+            ind_fused = bx1 * max_threads + tx1
+            with ib.if_scope(ind_fused < ind_full_range):
+                index_check = tir.LT(indices_ptr[ind_fused], tir.const(0, indices.dtype))
```

Comment:
Why can this block not be fused with the below block? Is it to prevent warp divergence?

Perhaps you can do something like:

`index = index + (index < 0) * axis_range`

If that is the case.

Reply:
Hello @AndrewZhaoLuo It was initially fused to the common loop. I decided that conditions is not good thing for GPU calculations aspecially wrapped to additional loops. But in this case it looks like the same (I can return it back if you think it is better way). Another thing is assert for probably shifted axis which is still out of bounds (theoretically we should do this check), but I do not know how I can do this check on ir builder side.

Reply:
It is very convenient and clear expression that you supposed. Is it the same as I wrote or use other tir procedures? I'm slightly aware about excess multiplication in it

Reply:
The arithmetic intensity per element loaded is very very low (~2) so I am not very worried about the multiplication. It's a pretty common trick to get "branch-less" programming.

Yes, I am not as familiar with IR Builder, I will say if you want to enforce indices being valid you can just do something like  modulo indices internally (for wrapping behavior). 

Personally I think it is ok to not check and assume caller will check and guarantee good inputs as we don't necessarily want the check in the base computation.

Reply:
Ok. I have fused block related to index shifting with general one and use your expression.
Code related to scatter_add was refactored in another branch and PR was prepared (see below), the latter is waiting for merging of this PR.